expression.go

```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package substraitgo
+
+import (
+	"fmt"
+
+	"github.com/substrait-io/substrait-go/proto"
+)
+
+func FuncArgFromProto(e *proto.FunctionArgument, baseSchema Type, ext ExtensionRegistry) (FuncArg, error) {
+	switch et := e.ArgType.(type) {
+	case *proto.FunctionArgument_Enum:
+		return Enum(et.Enum), nil
+	case *proto.FunctionArgument_Type:
+		return TypeFromProto(et.Type), nil
+	case *proto.FunctionArgument_Value:
+		return ExprFromProto(et.Value, baseSchema, ext)
+	}
+	return nil, ErrNotImplemented
+}
+
+func ExprFromProto(e *proto.Expression, baseSchema Type, ext ExtensionRegistry) (Expression, error) {
+	switch et := e.RexType.(type) {
+	case *proto.Expression_Literal_:
+		return LiteralFromProto(et.Literal), nil
+	case *proto.Expression_Selection:
+		return FieldReferenceFromProto(et.Selection, baseSchema, ext)
+	case *proto.Expression_ScalarFunction_:
+		var err error
+		args := make([]FuncArg, len(et.ScalarFunction.Arguments))
+		for i, a := range et.ScalarFunction.Arguments {
+			if args[i], err = FuncArgFromProto(a, baseSchema, ext); err != nil {
+				return nil, err
+			}
+		}
+
+		id, ok := ext.DecodeFunc(et.ScalarFunction.FunctionReference)
+		if !ok {
+			return nil, ErrNotFound
+		}
+
+		return &ScalarFunction{
+			FuncRef:    et.ScalarFunction.FunctionReference,
+			ID:         id,
+			Args:       args,
+			Options:    et.ScalarFunction.Options,
+			OutputType: TypeFromProto(et.ScalarFunction.OutputType),
+		}, nil
+	case *proto.Expression_WindowFunction_:
+		var err error
+		args := make([]FuncArg, len(et.WindowFunction.Arguments))
+		for i, a := range et.WindowFunction.Arguments {
+			if args[i], err = FuncArgFromProto(a, baseSchema, ext); err != nil {
+				return nil, err
+			}
+		}
+
+		parts := make([]Expression, len(et.WindowFunction.Partitions))
+		for i, p := range et.WindowFunction.Partitions {
+			if parts[i], err = ExprFromProto(p, baseSchema, ext); err != nil {
+				return nil, err
+			}
+		}
+
+		sorts := make([]SortField, len(et.WindowFunction.Sorts))
+		for i, s := range et.WindowFunction.Sorts {
+			if sorts[i], err = SortFieldFromProto(s, baseSchema, ext); err != nil {
+				return nil, err
+			}
+		}
+
+		id, ok := ext.DecodeFunc(et.WindowFunction.FunctionReference)
+		if !ok {
+			return nil, ErrNotFound
+		}
+
+		return &WindowFunction{
+			FuncRef:    et.WindowFunction.FunctionReference,
+			ID:         id,
+			Args:       args,
+			Options:    et.WindowFunction.Options,
+			OutputType: TypeFromProto(et.WindowFunction.OutputType),
+			Phase:      et.WindowFunction.Phase,
+			Invocation: et.WindowFunction.Invocation,
+			Partitions: parts,
+			Sorts:      sorts,
+			LowerBound: BoundFromProto(et.WindowFunction.LowerBound),
+			UpperBound: BoundFromProto(et.WindowFunction.UpperBound),
+		}, nil
+	case *proto.Expression_IfThen_:
```

Comment:
Any reason not to implement these?

Reply:
I was just trying to keep the size of this PR down by not implementing these yet, and instead implementing them in the next PR. If you'd prefer I can implement them now and add them to this PR.